File: packages/shared/src/dashboard-tiles.test.ts
import { describe, expect, it } from "vitest";

import {
  AGED_MILLIS,
  DASHBOARD_TILE_LABELS,
  dashboardTiles,
  type DashboardTileInput,
} from "./dashboard-tiles";

const MINUTE = 60 * 1000;
const HOUR = 60 * MINUTE;

/** Nothing behind on anything — the caught-up morning. */
const EMPTY: DashboardTileInput = {
  unassignedAgesMillis: [],
  waiting: [],
  tasks: [],
  unreadAgesMillis: [],
};

const ids = (input: DashboardTileInput) => dashboardTiles(input).map((t) => t.id);

describe("#540 the strip is ordered by what to do first", () => {
  it("leads with the tile that has something overdue", () => {
    // The complaint this fixes: the order never changed, so the most urgent
    // thing sat wherever its category happened to be. An overdue task is the
    // strongest signal there is, so it goes first even though "My tasks" is
    // third in the reading order below.
    expect(
      ids({
        ...EMPTY,
        unreadAgesMillis: [5 * MINUTE],
        tasks: [{ ageMillis: 2 * HOUR, overdue: true }],
      })[0],
    ).toBe("tasks");
  });

  it("puts an aged tile ahead of a busier fresh one", () => {
    // Twelve unread from five minutes ago is an ordinary morning. One thread
    // waiting since yesterday is a customer wondering if anybody read it. Count
    // is not urgency, which is the whole reason a bare number was not enough.
    const order = ids({
      ...EMPTY,
      unreadAgesMillis: Array.from({ length: 12 }, () => 5 * MINUTE),
      waiting: [{ ageMillis: 26 * HOUR, overdue: false }],
    });
    expect(order[0]).toBe("waiting");
  });

  it("keeps empty tiles, at the end, rather than removing them", () => {
    // A strip whose tiles come and go is one nobody can build a habit around —
    // and "nothing unassigned" is worth seeing. They keep their place; they just
    // stop competing for the front.
    const order = ids({ ...EMPTY, unreadAgesMillis: [10 * MINUTE] });
    expect(order).toHaveLength(4);
    expect(order[0]).toBe("unread");
    expect(order.slice(1).sort()).toEqual(["tasks", "unassigned", "waiting"]);
  });

  it("does not reshuffle over a few minutes' difference", () => {
    // The rank is coarse on purpose. If two tiles swapped every time one aged a
    // minute past the other, the strip would have moved every time somebody
    // looked at it, and an owner could never learn where anything is.
    const a = ids({
      ...EMPTY,
      waiting: [{ ageMillis: 30 * MINUTE, overdue: false }],
      unreadAgesMillis: [31 * MINUTE],
    });
    const b = ids({
      ...EMPTY,
      waiting: [{ ageMillis: 32 * MINUTE, overdue: false }],
      unreadAgesMillis: [31 * MINUTE],
    });
    // Both are "present but not aged", so both keep the page's reading order.
    expect(a.slice(0, 2)).toEqual(["waiting", "unread"]);
    expect(b.slice(0, 2)).toEqual(["waiting", "unread"]);
  });

  it("falls back to the order of the sections under it", () => {
    // With nothing to separate them the strip must agree with the page below,
    // because a strip that disagrees with what it sits on is worse than one that
    // never moves.
    expect(
      ids({
        unassignedAgesMillis: [MINUTE],
        waiting: [{ ageMillis: MINUTE, overdue: false }],
        tasks: [{ ageMillis: MINUTE, overdue: false }],
        unreadAgesMillis: [MINUTE],
      }),
    ).toEqual(["unassigned", "waiting", "tasks", "unread"]);
  });

  it("is stable when nothing at all is happening", () => {
    expect(ids(EMPTY)).toEqual(["unassigned", "waiting", "tasks", "unread"]);
  });
});

describe("#540 each tile carries a reason, not just a number", () => {
  it("reports how many are overdue when any are", () => {
    const tiles = dashboardTiles({
      ...EMPTY,
      tasks: [
        { ageMillis: 3 * HOUR, overdue: true },
        { ageMillis: 2 * HOUR, overdue: true },
        { ageMillis: MINUTE, overdue: false },
      ],
    });
    const tasks = tiles.find((t) => t.id === "tasks");
    expect(tasks?.count).toBe(3);
    expect(tasks?.signal).toEqual({ kind: "overdue", count: 2 });
  });

  it("reports the OLDEST age when nothing is overdue", () => {
    // The oldest, not the newest and not an average: the question a person is
    // asking is "how long has somebody been waiting on me", and the answer is
    // the worst case, not the typical one.
    const tiles = dashboardTiles({
      ...EMPTY,
      unreadAgesMillis: [10 * MINUTE, 3 * HOUR, MINUTE],
    });
    expect(tiles.find((t) => t.id === "unread")?.signal).toEqual({
      kind: "oldest",
      ageMillis: 3 * HOUR,
    });
  });

  it("says nothing for an empty tile rather than inventing a zero", () => {
    expect(dashboardTiles(EMPTY).every((t) => t.signal === null)).toBe(true);
  });

  it("never calls unassigned work overdue", () => {
    // Nobody owns it, so it cannot be late to a person. Age is the whole signal
    // — unclaimed work going stale is the failure, and calling it "overdue" would
    // imply somebody had already been asked.
    const tiles = dashboardTiles({
      ...EMPTY,
      unassignedAgesMillis: [40 * HOUR],
    });
    expect(tiles.find((t) => t.id === "unassigned")?.signal).toEqual({
      kind: "oldest",
      ageMillis: 40 * HOUR,
    });
  });

  it("ignores a task with no due date when counting overdue", () => {
    // A task nobody dated cannot be late. Counting it would put a red number on
    // a strip for something no one promised.
    const tiles = dashboardTiles({
      ...EMPTY,
      tasks: [
        { ageMillis: null, overdue: false },
        { ageMillis: 2 * HOUR, overdue: false },
      ],
    });
    const tasks = tiles.find((t) => t.id === "tasks");
    expect(tasks?.count).toBe(2);
    expect(tasks?.signal).toEqual({ kind: "oldest", ageMillis: 2 * HOUR });
  });

  it("treats four hours as the line between ordinary and too long", () => {
    // Stated as a test so the threshold is a decision rather than a magic number
    // somebody later tunes without knowing what it was for: this is the morning
    // triage question, not #388's five-minute reply window.
    expect(AGED_MILLIS).toBe(4 * HOUR);
    const justUnder = ids({
      ...EMPTY,
      waiting: [{ ageMillis: AGED_MILLIS - MINUTE, overdue: false }],
      unassignedAgesMillis: [MINUTE],
    });
    // Not aged yet, so the page's reading order holds.
    expect(justUnder[0]).toBe("unassigned");
    const justOver = ids({
      ...EMPTY,
      waiting: [{ ageMillis: AGED_MILLIS + MINUTE, overdue: false }],
      unassignedAgesMillis: [MINUTE],
    });
    expect(justOver[0]).toBe("waiting");
  });
});

describe("#540 every tile can be named", () => {
  it("has a label for each id the ordering can return", () => {
    for (const tile of dashboardTiles(EMPTY)) {
      expect(DASHBOARD_TILE_LABELS[tile.id]).toBeTruthy();
    }
  });
});

File: apps/android/app/src/test/kotlin/com/loonext/android/core/dashboard/DashboardTilesTest.kt
package com.loonext.android.core.dashboard

import java.io.File
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * #540 — the landing screen leads with the queue that needs doing first.
 *
 * Two halves. The behaviour tests below assert the rules; the parity test reads
 * the shared TypeScript, because this is a hand-port and nothing about Kotlin says
 * the original moved.
 */
class DashboardTilesTest {

    private val minute = 60_000L
    private val hour = 60 * minute

    private val empty = DashboardTiles.Input(
        unassignedAges = emptyList(),
        waiting = emptyList(),
        tasks = emptyList(),
        unreadAges = emptyList(),
    )

    private fun order(input: DashboardTiles.Input) =
        DashboardTiles.order(input).map { it.tile }

    @Test
    fun `an overdue task leads, whatever the reading order says`() {
        // The complaint this fixes: the order never changed, so the most urgent
        // thing sat wherever its section happened to be written.
        val result = order(
            empty.copy(
                unreadAges = listOf(5 * minute),
                tasks = listOf(DashboardTiles.Row(2 * hour, overdue = true)),
            ),
        )
        assertEquals(DashboardTiles.Tile.TASKS, result.first())
    }

    @Test
    fun `a stale queue beats a busier fresh one`() {
        // Count is not urgency. Twelve unread from five minutes ago is an ordinary
        // morning; one thread waiting since yesterday is a customer wondering
        // whether anybody read it.
        val result = order(
            empty.copy(
                unreadAges = List(12) { 5 * minute },
                waiting = listOf(DashboardTiles.Row(26 * hour, overdue = false)),
            ),
        )
        assertEquals(DashboardTiles.Tile.WAITING, result.first())
    }

    @Test
    fun `two fresh queues do not shuffle`() {
        // A minute between them must not swap them, or the screen has rearranged
        // every time somebody looked at it.
        val a = order(
            empty.copy(
                waiting = listOf(DashboardTiles.Row(30 * minute, overdue = false)),
                unreadAges = listOf(31 * minute),
            ),
        )
        val b = order(
            empty.copy(
                waiting = listOf(DashboardTiles.Row(32 * minute, overdue = false)),
                unreadAges = listOf(31 * minute),
            ),
        )
        assertEquals(listOf(DashboardTiles.Tile.WAITING, DashboardTiles.Tile.UNREAD), a.take(2))
        assertEquals(listOf(DashboardTiles.Tile.WAITING, DashboardTiles.Tile.UNREAD), b.take(2))
    }

    @Test
    fun `empty queues keep their place at the end`() {
        val result = order(empty.copy(unreadAges = listOf(10 * minute)))
        assertEquals(4, result.size)
        assertEquals(DashboardTiles.Tile.UNREAD, result.first())
    }

    @Test
    fun `nothing happening leaves the declared order alone`() {
        assertEquals(
            listOf(
                DashboardTiles.Tile.UNASSIGNED,
                DashboardTiles.Tile.WAITING,
                DashboardTiles.Tile.TASKS,
                DashboardTiles.Tile.UNREAD,
            ),
            order(empty),
        )
    }

    @Test
    fun `unassigned work is never called overdue`() {
        // Nobody owns it, so it cannot be late to a person.
        val entry = DashboardTiles.order(empty.copy(unassignedAges = listOf(40 * hour)))
            .first { it.tile == DashboardTiles.Tile.UNASSIGNED }
        assertEquals(DashboardTiles.Signal.Oldest(40 * hour), entry.signal)
    }

    @Test
    fun `a task with no due date cannot be overdue`() {
        val entry = DashboardTiles.order(
            empty.copy(
                tasks = listOf(
                    DashboardTiles.Row(null, overdue = false),
                    DashboardTiles.Row(2 * hour, overdue = false),
                ),
            ),
        ).first { it.tile == DashboardTiles.Tile.TASKS }
        assertEquals(2, entry.count)
        assertEquals(DashboardTiles.Signal.Oldest(2 * hour), entry.signal)
    }

    // ---------------------------------------------------- against the original

    private fun repoFile(relative: String): String {
        var dir: File? = File("").absoluteFile
        while (dir != null) {
            val candidate = File(dir, relative)
            if (candidate.exists()) return candidate.readText()
            dir = dir.parentFile
        }
        throw AssertionError("$relative not found from ${File("").absolutePath}")
    }

    /**
     * The four-hour line is the same number on both clients.
     *
     * It is the one value a crew would notice differing — the point at which the
     * screen decides a queue has gone stale — so it is read out of the shared
     * module rather than pinned twice and left to drift.
     */
    @Test
    fun `the aged threshold matches the shared module`() {
        val shared = repoFile("packages/shared/src/dashboard-tiles.ts")
        assertTrue(
            "AGED_MILLIS has drifted from packages/shared/src/dashboard-tiles.ts",
            shared.contains("export const AGED_MILLIS = 4 * 60 * 60 * 1000"),
        )
        assertEquals(4L * 60L * 60L * 1000L, DashboardTiles.AGED_MILLIS)
    }

    /** And the four queues are the same four, in the same declared order. */
    @Test
    fun `the tile ids match the shared module, in order`() {
        val shared = repoFile("packages/shared/src/dashboard-tiles.ts")
        val declared = Regex("""export type DashboardTileId =([^;]+);""")
            .find(shared)
            ?.groupValues
            ?.get(1)
            ?: throw AssertionError("DashboardTileId is no longer a union in the shared module")
        val ids = Regex(""""([a-z]+)"""").findAll(declared).map { it.groupValues[1] }.toList()
        assertEquals(listOf("unassigned", "waiting", "tasks", "unread"), ids)
        assertEquals(
            ids.map { it.uppercase() },
            order(empty).map { it.name },
        )
    }
}
